content/existential-containers.md

```diff
@@ -0,0 +1,206 @@
+---
+layout: sip
+permalink: /sips/:title.html
+stage: design
+status: submitted
+presip-thread: n/a
+title: SIP-NN - Existential Containers
+---
+
+**By: Dimi Racordon and Eugene Flesselle and Matt Bovel**
+
+## History
+
+| Date          | Version            |
+|---------------|--------------------|
+| Nov 25th 2024 | Initial Draft      |
+
+## Summary
+
+Type classes have become a well-established feature in the Scala ecosystem to escape some of the shortcomings of subtyping with respect to extensibility.
+Unfortunately, type classes do not support run-time polymorphism and dynamic dispatch, two features typically taken for granted in Scala.
+
+This SIP proposes a feature called *existential containers* to address this problem.
+An existential container wraps a value together with a witness of its conformance to one or several type classes into an object exposing the API defined by these type classes.
+
+## Motivation
+
+Type classes can address some of the well-known limitations of subtyping with respect to extensibility, such as the ability to extend existing data types with new behaviors [1].
+A type class describes the interface of a generic _concept_ as a set of requirements, expressed in the form of operations and associated types.
+These requirements can be implemented for a specific type, thereby specifying how this type _models_ the concept.
+The following illustrates:
+
+```scala
+import shapes.{Square, Hexagon}
+
+trait TypeClass:
+  type Self
+
+trait Polygon extends TypeClass:
+  extension (self: Self)
+    def area: Double
+
+given Square is Polygon: ...
+given Hexagon is Polygon: ...
+```
+
+Defining `Polygon` as a type class rather than an abstract class to be inherited allows us to retroactively state that squares are polygons without modifying the definition of `Square`.
+Sticking to subtyping would require the definition of an inefficient and verbose wrapper class.
+
+Alas, type classes offer limited support for eliding type information at compile-time.
+Hence, it is difficult to manipulate heterogeneous collections or write procedures returning arbitrary values known to model a particular concept.
+The following illustrates:
+
+```scala
+def largest[T: Polygon](xs: Seq[T]): Option[T] =
+  xs.maxByOption(_.area)
+
+largest(List(Square(), Hexagon()))
+// error: No given instance of type Polygon{type Self = Square | Hex} was found for a context parameter of method largest
+```
+
+The call to `largest` is illegal because, although there exist witnesses of the `Polygon` and `Hexagon`'s conformance to `Polygon`, no such witness exists for their least common supertype.
+In other words, it is impossible to call `largest` with an heterogeneous sequence of polygons.
+
+## Proposed solution
```

Comment:
This section is under-specified. Reading it, I have no idea what the proposed solution is. 

* Probably a lot of the text from `Specification` should actually be up here
* We need several more use case examples, which are less contrived than `Square` `Rectangle` and `Polygon`

Someone should be able to read the proposed solution have _zero idea how it is implemented_, and still get the general idea of what the proposed language feature is about. Right now, that is not the case, so the proposal is incomplete

Reply:
I can write other examples but I disagree that the one in the document is contrived. It also generalizes very easily to any situation involving a set of classes conforming to the same type class. Any instance of a situation where the "upper bound" of a set of types is defined retroactively would look like the proposed example.

```
def lookup(key: List[Containing[Hashable]]) = ???
def zipAll[E](xss: List[Containing[Iterator]{ type Element = E }]) = ???
```

Instead of justifying existential containers, however, I propose to clarify that the SIP is about sugaring the selection of a method rather than proposing the containers themselves, as those can be defined in library space.

Reply:
You really need to justify both the language changes and library support. The language change is useless without the library support, but most people would not be familiar with this library-level technique.

`List[Containing[Hashable]]` is definitely a lot more concrete than `Seq[Containing[Polygon]]`. We need more examples like that. 

* I've found use for the equivalent of `Seq[Containing[upickle.default.Writer]]` myself, which contains a list of things that can be converted to JSON

* Scalatags currently uses methods taking `(args: Frag*)` with implicit conversions from various types to `Frag`. it could conceivable instead use `(args: Containing[Fragable]*)` with a `Fragable` typeclass to render the value to HTML. Is that better or worse? Why?